code/datums/components/proximity_monitor.dm

```diff
@@ -1,139 +1,474 @@
+ */
+/datum/component/proximity_monitor/advanced
+	name = "Advanced energy field"
+	field_checker_type = /obj/effect/abstract/proximity_checker/advanced/inner_field
+	/// The type of checker object that should be used for the field edges.
+	var/edge_checker_type = /obj/effect/abstract/proximity_checker/advanced/edge_field
+	/// Make a distinction between edge checkers and field checkers seperately.
+	var/uses_edge_checkers = FALSE
+	/// Do any of the proximity_checker objects need to process things sitting on their tile?
+	var/requires_processing = FALSE
+	/// Should the main field checkers process things on their tile?
+	var/process_field_checkers = FALSE
+	/// Should the edge field checkers process things on their tile?
+	var/process_edge_checkers = FALSE
+	/// A list of proximity_checkers in the inner field. Excludes checkers on the edge of the field.
+	var/list/field_checkers
+	/// A list of proximity_checkers on the edge of the field.
+	var/list/edge_checkers
+
+/datum/component/proximity_monitor/advanced/Initialize(_radius = 1, _always_active = FALSE)
+	. = ..()
+	if(requires_processing)
+		START_PROCESSING(SSfields, src)
+
+/datum/component/proximity_monitor/advanced/Destroy(force, silent)
+	STOP_PROCESSING(SSfields, src)
+	QDEL_NULL(field_checkers)
+	QDEL_NULL(edge_checkers)
+	return ..()
+
+/datum/component/proximity_monitor/advanced/create_prox_checkers()
+	if(!uses_edge_checkers)
+		..() // We don't need to make a distinction between field and edge checkers, use the parent.
+		if(process_field_checkers)
+			field_checkers = proximity_checkers.Copy() // Still allows for field checkers to use processing.
+		return
+
+	LAZYINITLIST(proximity_checkers)
+	LAZYINITLIST(field_checkers)
+	LAZYINITLIST(edge_checkers)
+
+	var/turf/parent_turf = get_turf(parent)
+	for(var/T in RANGE_TURFS(radius, parent_turf))
+		if(get_dist(T, parent_turf) == radius)
+			edge_checkers += create_single_prox_checker(T, edge_checker_type)
+			continue
+		field_checkers += create_single_prox_checker(T)
+
+/datum/component/proximity_monitor/advanced/recenter_prox_checkers()
+	if(!uses_edge_checkers)
+		return ..() // We don't need to make a distinction between field and edge checkers, use the parent.
+
+	var/turf/parent_turf = get_turf(parent)
+	var/inner_index = 1
+	var/edge_index = 1
+
+	for(var/T in RANGE_TURFS(radius, parent_turf))
+		var/obj/checker
+		if(get_dist(T, parent_turf) == radius) // If it's at this distance, it's on the edge of the field.
+			checker = edge_checkers[edge_index++]
+			checker.loc = T
+			continue
+		checker = field_checkers[inner_index++]
+		checker.loc = T
+
+/datum/component/proximity_monitor/advanced/process()
+	if(process_field_checkers)
+		for(var/checker in field_checkers)
+			var/obj/effect/abstract/proximity_checker/advanced/inner_field/F = checker
+			process_inner_checker(F)
```

Comment:
```suggestion
			process_inner_checker(checker)
```

Probably fine to pass as is without casting

Reply:
Uh, yeah, I don't know why I did that